akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -22,6 +25,85 @@ object ORMap {
    */
   def unapply[A, B <: ReplicatedData](m: ORMap[A, B]): Option[Map[A, B]] = Some(m.entries)
 
+  sealed trait DeltaOp extends ReplicatedDelta with RequiresCausalDeliveryOfDeltas {
+    type T = DeltaOp
+    override def zero: DeltaReplicatedData
+  }
+
+  private[akka] final val VanillaORMapTag = 0
+  private[akka] final val PNCounterMapTag = 1
+  private[akka] final val ORMultiMapTag = 2
+  private[akka] final val ORMultiMapDeltaDeltaTag = 3
+  private[akka] final val LWWMapTag = 4
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def tagDelta(op: DeltaOp)(tag: Int) = op match {
+    case p: PutDeltaOp[_, _]       ⇒ p.copy(zeroTag = tag)
+    case u: UpdateDeltaOp[_, _]    ⇒ u.copy(zeroTag = tag)
+    case r: RemoveDeltaOp[_, _]    ⇒ r.copy(zeroTag = tag)
+    case k: RemoveKeyDeltaOp[_, _] ⇒ k.copy(zeroTag = tag)
+    case DeltaGroup(ops) ⇒ DeltaGroup(ops.map {
+      case p: PutDeltaOp[_, _]       ⇒ p.copy(zeroTag = tag)
+      case u: UpdateDeltaOp[_, _]    ⇒ u.copy(zeroTag = tag)
+      case r: RemoveDeltaOp[_, _]    ⇒ r.copy(zeroTag = tag)
+      case k: RemoveKeyDeltaOp[_, _] ⇒ k.copy(zeroTag = tag)
+      case _: DeltaGroup[_, _]       ⇒ throw new IllegalArgumentException("ORMap.DeltaGroup should not be nested")
+    })
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] sealed abstract class AtomicDeltaOp[A, B <: ReplicatedData] extends DeltaOp {
+    def underlying: ORSet.DeltaOp
+    def value: Map[A, B]
+    def zeroTag: Int
+    override def zero: DeltaReplicatedData = zeroTag match {
+      case PNCounterMapTag         ⇒ PNCounterMap.empty[A]
```

Comment:
we could get rid of these matches by making `zeroTag` something richer than an `Int`, such as
```
trait ZeroTag {
  def zero: DeltaReplicatedData
  def value: Int
}

case object PNCounterMapTag extends ZeroTag {
  override def zero: DeltaReplicatedData = PNCounterMap.empty
  override def value: Int = 1
}
```

Another advantage of this is that it would be easier to open it up for custom implementations (outside akka) if needed.

The serialized format would still just be `Int` and match in the deserialization to pick the right instance.